examples/simple-router/simple-router.rb

```diff
@@ -0,0 +1,167 @@
+require "pio"
+
+require "arp-table"
+require "interface"
+require "routing-table"
+require "router-utils"
+
+class SimpleRouter < Trema::Controller
+  include RouterUtils
+
+  def start( _args )
+    load "simple_router.conf"
+    @interfaces = Interfaces.new( $interface )
+    @arp_table = ARPTable.new
+    @routing_table = RoutingTable.new( $route )
+  end
+
+  def switch_ready( datapath_id )
+    logger.info "#{datapath_id.to_hex} is connected"
+    @switche = datapath_id
+    init_flows( datapath_id )
+  end
+
+  def packet_in( datapath_id, message )
+    return if not to_me?( message )
+
+    case message.data
+    when Pio::Arp::Request
+      handle_arp_request( datapath_id, message )
+    when Pio::Arp::Reply
+      handle_arp_reply( message )
+    when Pio::Parser::IPv4Packet
+      handle_ipv4( datapath_id, message )
+    else
+      fail 'Failed to handle packet_in data type'
+    end
+  end
+
+  private
+
+  def init_flows( datapath_id )
+    send_flow_mod_delete( datapath_id, match: Match.new() )
+  end
+
+  def to_me?( message )
+    return true if message.destination_mac.broadcast?
+
+    interface = @interfaces.find_by_port( message.in_port )
+    if interface and interface.has?( message.destination_mac )
+      return true
+    end
+  end
+
+  def handle_arp_request( datapath_id, message )
+    port = message.in_port
+    destination_mac = message.data.target_protocol_address
+    interface = @interfaces.find_by_port_and_ipaddr( port, destination_mac )
+    if interface
+      arp_reply = create_arp_reply_from( message, interface.hwaddr )
+      packet_out_raw( datapath_id, arp_reply, SendOutPort.new( interface.port ) )
+    end
+  end
+
+  def handle_arp_reply( message )
+    @arp_table.update( message.in_port,
+                      message.sender_protocol_address,
+                      message.source_mac
+                     )
+  end
+
+  def handle_ipv4( datapath_id, message )
+    if should_forward?( message )
+      forward( datapath_id, message )
+    elsif message.data.ip_protocol == 1
+      icmp = Pio::Icmp.read( message.raw_data )
+      if icmp.icmp_type == 8
+        handle_icmpv4_echo_request( datapath_id, message )
+      else
+        fail "Faild to handle icmp type"
```

Comment:
Prefer single-quoted strings when you don't need string interpolation or special symbols.